gpbft/gpbft.go

```diff
@@ -467,11 +467,23 @@ func (i *instance) tryCurrentPhase() error {
 	}
 }
 
-// Checks message validity, including justification and signatures.
-// An invalid message can never become valid, so may be dropped.
-// This is a pure function and does not modify its arguments.
-// It must be safe for concurrent use.
-func ValidateMessage(powerTable *PowerTable, beacon []byte, host Host, msg *GMessage) error {
+// CachedJustificationChecker checks whether a justification at an instance is
+// cached, and if so returns true. The justification must be provided as CBOR
+// encoding of the entire Justification.
+//
+// Implementations must be safe for concurrent use.
+type CachedJustificationChecker func(instance uint64, namespace, jcbor []byte) (bool, error)
+
+// ValidateMessage checks message validity, including justification and
+// signatures. An invalid message can never become valid, so may be dropped. This
+// is a pure function and does not modify its arguments. It must be safe for
+// concurrent use.
+//
+// The caller may optionally specify CachedJustificationChecker to shortcut
+// validation of justifications. If the checker is non-nil, then the validation
+// will check if the justification is present and if so considers the
+// justification valid. Otherwise, it fully validates the justification.
+func ValidateMessage(powerTable *PowerTable, beacon []byte, host Host, msg *GMessage, isJustificationCached CachedJustificationChecker) error {
```

Comment:
This is getting ugly (and was already ugly due to the number of arguments). Let's just inline it in `Participant.ValidateMessage`, that lets us keep all the caching in one place.

Reply:
it is indeed ugly; will do. 🙏